File: src/frontends/tensorflow_common/src/op/log1p.cpp
// Copyright (C) 2018-2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//

#include "common_op_table.hpp"
#include "openvino/op/add.hpp"
#include "openvino/op/log.hpp"
#include "utils.hpp"

using namespace std;
using namespace ov::op;

namespace ov {
namespace frontend {
namespace tensorflow {
namespace op {

OutputVector translate_log_1p_op(const NodeContext& node) {
    // compute element-wise natural logarithm of (1 + x),
    // where x - input
    default_op_checks(node, 1, {"Log1p"});
    auto x = node.get_input(0);
    auto const_one = create_same_type_const_scalar<float>(x, 1);
    auto x_plus_one = make_shared<v1::Add>(x, const_one);
    auto log1p = make_shared<v0::Log>(x_plus_one);
    set_node_name(node.get_name(), log1p);
    return {log1p};
}
}  // namespace op
}  // namespace tensorflow
}  // namespace frontend
}  // namespace ov

File: tools/mo/openvino/tools/mo/front/Log1p.py
# Copyright (C) 2018-2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import numpy as np

from openvino.tools.mo.ops.activation_ops import Log
from openvino.tools.mo.ops.elementwise import Add
from openvino.tools.mo.front.common.partial_infer.utils import mo_array
from openvino.tools.mo.front.common.replacement import FrontReplacementOp
from openvino.tools.mo.graph.graph import Graph, Node, rename_nodes
from openvino.tools.mo.ops.const import Const


class Log1p(FrontReplacementOp):
    """
    Log1p computes natural logarithm of (1 + x) element-wise.
    It replaces Log1p operation with Add -> Log.
    """
    op = "Log1p"
    enabled = True

    def replace_op(self, graph: Graph, node: Node):
        node_name = node.soft_get('name', node.id)
        const_dtype = np.float32
        if node.has_valid('data_type'):
            const_dtype = node.data_type
        const = Const(graph, {'value': mo_array([1], dtype=const_dtype)}).create_node()
        add = Add(graph, {'name': node.name + '/Add_'}).create_node()
        log = Log(graph, {'name': node.name + '/Log_'}).create_node()

        # Connect nodes: input -> Add -> Log
        const.out_port(0).connect(add.in_port(0))
        node.in_port(0).get_connection().set_destination(add.in_port(1))
        add.out_port(0).connect(log.in_port(0))
        rename_nodes([(node, node_name + '/delete'), (log, node_name)])

        # The "explicit" version of the return value is: [(out_node.id, 0)])
        return [log.id]
